src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,78 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
-    private static final Injector injector = new Injector();
+    private static final Injector INJECTOR = new Injector();
+    private static final Map<Class<?>, Class<?>> INTERFACE_IMPLS = new HashMap<>();
+    private static final Map<Class<?>, Object> INSTANCES = new HashMap<>();
 
     public static Injector getInjector() {
-        return injector;
+        return INJECTOR;
+    }
+
+    static {
+        INTERFACE_IMPLS.put(FileReaderService.class, FileReaderServiceImpl.class);
+        INTERFACE_IMPLS.put(ProductParser.class, ProductParserImpl.class);
+        INTERFACE_IMPLS.put(ProductService.class, ProductServiceImpl.class);
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        if (!interfaceClazz.isAnnotationPresent(Component.class)) {
+            throw new RuntimeException("This class has no annotation Component");
+        }
+        Class<?> clazz = findImplementation(interfaceClazz);
+        Field[] declaredFields = clazz.getDeclaredFields();
+        Object clazzImplInstance = null;
+        for (Field field : declaredFields) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                Object fieldInstance = getInstance(field.getType());
+                clazzImplInstance = createNewInstance(clazz);
+
+                field.setAccessible(true);
+                try {
+                    field.set(clazzImplInstance, fieldInstance);
+                } catch (IllegalAccessException e) {
+                    throw new RuntimeException("Can't initialize field value. "
+                            + "Class: "
+                            + clazz.getName()
+                            + ". Field: "
+                            + field.getName());
+                }
+            }
+        }
+        if (clazzImplInstance == null) {
+            clazzImplInstance = createNewInstance(clazz);
+        }
+        return clazzImplInstance;
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (INSTANCES.containsKey(clazz)) {
+            return INSTANCES.get(clazz);
+        }
+        try {
+            Constructor<?> constructor = clazz.getConstructor();
+            Object instance = constructor.newInstance();
+            INSTANCES.put(clazz, instance);
+            return instance;
+        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException
```

Comment:
As states in common mistakes - better change multiple exception with one parent - ReflectiveOperationException.